### DISCUSSION TITLE
Logistic Regression Classification model

### DISCUSSION THREAD

**Discussion by memora0101** [explosion/spaCy#6573] [Help: Coding & Implementations]
## Your Environment
import pandas as pd
import string
import spacy
import numpy as np
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk.tokenize import TweetTokenizer
from collections import Counter
from spacy.lang.en import English
from spacy.lang.en.stop_words import STOP_WORDS
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer
from sklearn.base import TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC 
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics
from sklearn import preprocessing

df1=pd.read_csv("Benchmark.csv",encoding='unicode_escape')
df1.head(10)
**output:**
	ï»¿chatid	user	chatorder	text	                                                          ANGRY	HAPPY
0	1	Robert	1	Hello, is this apple product support?	                           0.0	   0.3
1	1	Bot	        2	Yes, it is. How may I help you?	                                    0.0	    0.7
2	1	Robert	3	I have a problem with one of the products that...	    0.7	   0.0
3	1	Bot	        4	I am sorry for that. Can you please tell which...	            0.0	   0.0
4	1	Robert	5	It is an iphone 10, it has been a week about n...             1.0	   0.0
5	1	Bot	        6	I understand the situation and I am glad to as...	     0.0	   2.0
6	1	Robert	7	My order id is 54569723 and I would like to ha...	    0.3	   0.0
7	1	Bot	        8	A replacement request has been created, please...	    0.0	   0.7
8	1	Robert	9	Thank you for you help, have a good day	                    0.0	   4.3
9	1	Bot	        10	You're welcome, have a great day!	                            0.0	   4.7

df1.info()
**output:**
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197 entries, 0 to 196
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   ï»¿chatid   197 non-null    int64  
 1   user            197 non-null    object 
 2   chatorder  197 non-null    int64  
 3   text            197 non-null    object 
 4   ANGRY      197 non-null    float64
 5   HAPPY      197 non-null    float64
dtypes: float64(2), int64(2), object(2)
memory usage: 9.4+ KB

def binary_classifyer(df1):
    if df1['ANGRY']>df1['HAPPY']:
        return 0
    elif df1['ANGRY']<df1['HAPPY']:
        return 1
    elif df1['ANGRY']>.5:
        return 0
    elif df1['ANGRY']<.5:
        return 1  
df1['classification']=df1.apply(binary_classifyer, axis=1)
df1.head()

**output:**
  | ï»¿chatid | user | chatorder | text                                                                        | ANGRY | HAPPY | classification

1                   | Robert | 1            | Hello, is this apple product support?                     | 0.0     | 0.3       | 1
1                   | Bot       | 2           | Yes, it is. How may I help you?                               | 0.0      | 0.7       | 1
1                   | Robert | 3           | I have a problem with one of the products that... | 0.7      | 0.0       | 0
1                   | Bot       | 4           | I am sorry for that. Can you please tell which...   | 0.0      | 0.0       | 1
1                   | Robert | 5           | It is an iphone 10, it has been a week about n...   | 1.0       | 0.0       | 0

df1.classification.value_counts()

**output:**
1    126
0     71
Name: classification, dtype: int64

positive_text=df1.text[df1.classification==1]
negative_text=df1.text[df1.classification==0]
test_pos = positive_text[101:]
train_pos = positive_text[:101]
test_neg = negative_text[57:]
train_neg = negative_text[:57]

train_x = train_pos + train_neg 
test_x = test_pos + test_neg

train_y = np.append(np.ones((len(train_pos), 1)), np.zeros((len(train_neg), 1)), axis=0)
test_y = np.append(np.ones((len(test_pos), 1)), np.zeros((len(test_neg), 1)), axis=0)
print("train_y.shape = " + str(train_y.shape))
print("test_y.shape = " + str(test_y.shape))

**output:**
train_y.shape = (158, 1)
test_y.shape = (39, 1)

def process_text(wd):
    """Process tweet function.
    Input:
        wd: a string containing a sentence
    Output:
        sent_clean: a list of words containing the processed sentence
    """
    sent_clean=nlp(wd)
    sent_clean=[d for d in sent_clean if d not in STOP_WORDS]
    sent_clean=[wod.lemma_.lower() for wod in sent_clean if wod.is_alpha]

    return str(sent_clean)

print(df1['text'].apply(process_text))
**output:**
0      ['hello', 'be', 'this', 'apple', 'product', 's...
1      ['yes', '-pron-', 'be', 'how', 'may', '-pron-'...
2      ['-pron-', 'have', 'a', 'problem', 'with', 'on...
3      ['-pron-', 'be', 'sorry', 'for', 'that', 'can'...
4      ['-pron-', 'be', 'an', 'iphone', '-pron-', 'ha...
                             ...                        
192    ['connect', 'me', 'to', 'someone', 'else', 'now']
193    ['-pron-', 'will', 'not', 'apologize', 'to', '...
194        ['are', 'you', 'transfer', 'me', 'or', 'not']
195    ['may', '-pron-', 'have', '-pron-', 'phone', '...
196    ['-pron-', 'am', 'post', 'this', 'conversation...
Name: text, Length: 197, dtype: object

def build_freqs(sent, ys):
    """Build frequencies.
    Input:
        tweets: a list of sentence
        ys: an m x 1 array with the sentiment label of each sentence
            (either 0 or 1)
    Output:
        freqs: a dictionary mapping each (word, sentiment) pair to its
        frequency
    """
    # Convert np array to list since zip needs an iterable.
    # The squeeze is necessary or the list ends up with one element.
    # Also note that this is just a NOP if ys is already a list.
    yslist = np.squeeze(ys).tolist()

    # Start with an empty dictionary and populate it by looping over all sentences
    # and over all processed words in each sentence.
    freqs = {}
    for y, sent in zip(yslist, sent):
        for word in process_text (sent):
            pair = (word, y)
            if pair in freqs:
                freqs[pair] += 1
            else:
                freqs[pair] = 1

    return freqs

# create frequency dictionary
freqs = build_freqs(train_x, train_y)
# check the output
print("type(freqs) = " + str(type(freqs)))
print("len(freqs) = " + str(len(freqs.keys())))

**output:**
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-45-cb5b90c507a1> in <module>
      1 # create frequency dictionary
----> 2 freqs = build_freqs(train_x, train_y)
      3 # check the output
      4 print("type(freqs) = " + str(type(freqs)))
      5 print("len(freqs) = " + str(len(freqs.keys())))

<ipython-input-44-75bc1a2dfbaf> in build_freqs(sent, ys)
     18     freqs = {}
     19     for y, sent in zip(yslist, sent):
---> 20         for word in process_text (sent):
     21             pair = (word, y)
     22             if pair in freqs:

<ipython-input-38-827c8bb21570> in process_text(wd)
      7 
      8     """
----> 9     sent_clean=nlp(wd)
     10     sent_clean=[d for d in sent_clean if d not in STOP_WORDS]
     11     sent_clean=[wod.lemma_.lower() for wod in sent_clean if wod.is_alpha]

~/opt/anaconda3/lib/python3.7/site-packages/spacy/language.py in __call__(self, text, disable, component_cfg)
    435         DOCS: https://spacy.io/api/language#call
    436         """
--> 437         if len(text) > self.max_length:
    438             raise ValueError(
    439                 Errors.E088.format(length=len(text), max_length=self.max_length)

TypeError: object of type 'float' has no len()

spacy 2.1.8
- Operating System: IOS/Mac/apple
- Python Version Used:3.7.6
- spaCy Version Used:2.1.8
- Environment Information: I am running some errors into the following code above. I am working with a chatbot classifier project for my school and need the most of help figuring out if spacy is causing the TypeError

**Comment by svlandeg** [comment]
Hi! 

I moved your issue to the GH Discussions board, as we try to keep the Issue Tracker focused on specific bug reports. For a bug report, we'd really need a minimal code snippet that shows the issue with spaCy.

I find your post very difficult to read. Could you revise it and ensure you're using proper [markdown formatting](https://guides.github.com/features/mastering-markdown/), especially for the code bits & output quotes?
